File: LC104_MaximumDepthofBinaryTree/DFS/dfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    // DFS
    // Time Complexity: O(n)
    // Space Complexity: O(n)
public:
    int maxDepth(TreeNode* root) {
        if(root == nullptr) return 0;
        return dfs(root);
    }

    int dfs(TreeNode* node){
        //Pre
        if(node == nullptr){
            return 0;
        }
        int lLeft = dfs(node->left);
        int lRight = dfs(node->right);
        // Post
        return max(lLeft, lRight) + 1;
    }
};

File: LC104_MaximumDepthofBinaryTree/DFS/dfs.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # DFS
    # Time Complexty: O(n)
    # Space Complexity: O(n)
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return self.dfs(root)
    
    def dfs(self, node):
        lLeft = 0
        lRight = 0
        if node.left is not None:
            self.dfs(node.left)
        if node.right is not None:
            self.dfs(node.right)
        return max(lLeft, lRight) + 1